hoomd/GSDDumpWriter.cc

```diff
@@ -150,85 +299,61 @@ void GSDDumpWriter::analyze(uint64_t timestep)
     {
     Analyzer::analyze(timestep);
     int retval;
-    bool root = true;
-
-    // take particle data snapshot
-    m_exec_conf->msg->notice(10) << "GSD: taking particle data snapshot" << endl;
-    SnapshotParticleData<float> snapshot;
-    const std::map<unsigned int, unsigned int>& map = m_pdata->takeSnapshot<float>(snapshot);
-
-#ifdef ENABLE_MPI
-    // if we are not the root processor, do not perform file I/O
-    root = m_exec_conf->isRoot();
-#endif
 
     // open the file if it is not yet opened
-    if (!m_is_initialized && root)
+    if (!m_is_initialized)
         initFileIO();
 
     // truncate the file if requested
-    if (m_truncate && root)
+    if (m_truncate)
         {
-        m_exec_conf->msg->notice(10) << "GSD: truncating file" << endl;
-        retval = gsd_truncate(&m_handle);
-        GSDUtils::checkError(retval, m_fname);
-        }
+        if (m_exec_conf->isRoot())
+            {
+            m_exec_conf->msg->notice(10) << "GSD: truncating file" << endl;
+            retval = gsd_truncate(&m_handle);
+            GSDUtils::checkError(retval, m_fname);
+            }
 
-    uint64_t nframes = 0;
-    if (root)
-        {
-        nframes = gsd_get_nframes(&m_handle);
-        m_exec_conf->msg->notice(10)
-            << "GSD: " << m_fname << " has " << nframes << " frames" << endl;
+        m_nframes = 0;
         }
 
-#ifdef ENABLE_MPI
-    bcast(nframes, 0, m_exec_conf->getMPICommunicator());
-#endif
+    populateLocalFrame(m_local_frame, timestep);
 
-    if (root)
+#ifdef ENABLE_MPI
+    if (m_sysdef->isDomainDecomposed())
         {
-        // write out the frame header on all frames
-        writeFrameHeader(timestep);
+        gatherGlobalFrame(m_local_frame);
 
-        // only write out data chunk categories if requested, or if on frame 0
-        if (m_write_attribute || nframes == 0)
-            writeAttributes(snapshot, map);
-        if (m_write_property || nframes == 0)
-            writeProperties(snapshot, map);
-        if (m_write_momentum || nframes == 0)
-            writeMomenta(snapshot, map);
+        if (m_exec_conf->isRoot())
+            {
+            writeFrameHeader(m_global_frame);
+            writeAttributes(m_global_frame);
+            writeProperties(m_global_frame);
+            writeMomenta(m_global_frame);
+            }
+        }
+    else
+#endif
+        {
+        writeFrameHeader(m_local_frame);
+        writeAttributes(m_local_frame);
+        writeProperties(m_local_frame);
+        writeMomenta(m_local_frame);
         }
 
     // topology is only meaningful if this is the all group
     if (m_group->getNumMembersGlobal() == m_pdata->getNGlobal()
-        && (m_write_topology || nframes == 0))
+        && (m_write_topology || m_nframes == 0))
```

Comment:
It should not be an error to write the first frame of the file with a filter that selects a subset. Otherwise, users could never select a subset.

Do you think it would be useful for users to receive an exception on the 2nd frame written to the file if they both select a subset and include `'topology'` in the dynamic list?

Reply:
I was thinking of cases where users specify to write out `topology` that is explicitly in `dynamic`, but the filter excludes some particles. So only erroring when `m_write_topology == true` and `m_group->getNumMembersGlobal() != m_pdata->getNGlobal()`

Reply:
I don't think the error can be checked in `setDynamic`. The user may provide group that changes membership from one frame to the next.